### PR TITLE
Add Dilithium

### LINKED ISSUE #136: Add Round 2 signature scheme Dilithium

### PR DESCRIPTION
**Pull request by mkannwischer**
Resolves #136

#### Manually checked properties

* [x] No stringification macros
* [ ] Output-parameter pointers in functions are on the left
* [ ] Negative return values on failure of API functions (within restrictions of FO transform).
* [ ] `const` arguments are labeled as `const`
* [ ] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by mkannwischer** [comment]
This is ready for review. 
The changes compared to the reference version are quite significant; especially because it has to support the detached signature API which requires to use incremental hashing. 
We might want to reduce the number of casts in `rounding.c` - but this is the implicit casting in the reference implementation.

**Comment by mkannwischer** [comment]
Conflicts with #175.

File: crypto_sign/dilithium2/clean/rounding.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    return (a1 - 1) & 0xF;
+
+    /* If decompose does not divide out ALPHA:
+    if(hint == 0)
+      return a1;
+    else if(a0 > Q)
+      return (a1 + ALPHA) % (Q - 1);
+    else
+      return (a1 - ALPHA) % (Q - 1);
+    */
```
Comment (thomwiggers) [inline crypto_sign/dilithium2/clean/rounding.c]: what's this doing here? I'd argue commented-out code should go if it doesn't explain anything.

File: crypto_sign/dilithium2/clean/symmetric.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+} keccak_state;
+
+void PQCLEAN_DILITHIUM2_CLEAN_shake256_absorb(keccak_state *state, const uint8_t *input, size_t inlen);
+void PQCLEAN_DILITHIUM2_CLEAN_shake256_squeezeblocks(uint8_t *output, size_t nblocks, keccak_state *state);
+void PQCLEAN_DILITHIUM2_CLEAN_shake128_absorb(keccak_state *state, const uint8_t *input, size_t inlen);
+void PQCLEAN_DILITHIUM2_CLEAN_shake128_squeezeblocks(uint8_t *output, size_t nblocks, keccak_state *state);
+#define PQCLEAN_DILITHIUM2_CLEAN_shake256_absorb(STATE, IN, INBYTES) (shake256_absorb((uint64_t *) (STATE), IN, INBYTES))
+#define PQCLEAN_DILITHIUM2_CLEAN_shake256_squeezeblocks(OUTPUT, OUTBLOCKS, STATE) (shake256_squeezeblocks(OUTPUT, OUTBLOCKS, (uint64_t *) (STATE)))
+#define PQCLEAN_DILITHIUM2_CLEAN_shake128_absorb(STATE, IN, INBYTES) (shake128_absorb((uint64_t *) (STATE), IN, INBYTES))
+#define PQCLEAN_DILITHIUM2_CLEAN_shake128_squeezeblocks(OUTPUT, OUTBLOCKS, STATE) (shake128_squeezeblocks(OUTPUT, OUTBLOCKS, (uint64_t *) (STATE)))
```
Comment (thomwiggers) [inline crypto_sign/dilithium2/clean/symmetric.h]: do you still need these to be defined if you reimplement `shake256`  in terms of #175 ?

Comment (mkannwischer) [reply crypto_sign/dilithium2/clean/symmetric.h]: nope - I removed those.

**Comment by mkannwischer** [comment]
Thanks @thomwiggers  for the review - I've resolved those comments. 
I've also ported it to the new #175  fips202 structs.

**Comment by mkannwischer** [comment]
@cryptojedi, can we get this merged soon? I'd like to have it in pqm4 for the 31 May deadline (that still requires some work).

**Comment by dstebila** [comment]
> @cryptojedi, can we get this merged soon? I'd like to have it in pqm4 for the 31 May deadline (that still requires some work).

Anything you're waiting for to merge? If not, go ahead, there's already been a review from Thom.

**Comment by mkannwischer** [comment]
> > @cryptojedi, can we get this merged soon? I'd like to have it in pqm4 for the 31 May deadline (that still requires some work).
> 
> Anything you're waiting for to merge? If not, go ahead, there's already been a review from Thom.

I'd like some feedback from the Dilithium team, especially on the detached signature implementation. 
I'm in contact with "the Dilithium team" and it shouldn't take much longer :)

**Comment by mkannwischer** [comment]
Ok, since I've received two requests of people who want to use Dilithium in pqm4 now, I'll merge this. 
In case anything needs to be changed, please let me know.

**Pull request merged**